libfwupd/fwupd-client.c

```diff
@@ -2039,6 +2112,129 @@ fwupd_client_get_remote_by_id (FwupdClient *client,
 	return g_object_ref (remote);
 }
 
+/**
+ * fwupd_client_set_user_agent:
+ * @client: A #FwupdClient
+ * @user_agent: the user agent ID, e.g. `gnome-software/3.34.1`
+ *
+ * Sets the user agent that is used for downloading.
+ *
+ * Since: 1.5.0
+ **/
+void
+fwupd_client_set_user_agent (FwupdClient *client, const gchar *user_agent)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (client);
+	g_return_if_fail (FWUPD_IS_CLIENT (client));
+	g_return_if_fail (user_agent != NULL);
+	g_free (priv->user_agent);
+	priv->user_agent = g_strdup (user_agent);
+}
+
+static void
+fwupd_client_download_chunk_cb (SoupMessage *msg, SoupBuffer *chunk, gpointer user_data)
+{
+	guint percentage;
+	goffset header_size;
+	goffset body_length;
+	FwupdClient *client = FWUPD_CLIENT (user_data);
+
+	/* if it's returning "Found" or an error, ignore the percentage */
+	if (msg->status_code != SOUP_STATUS_OK) {
+		g_debug ("ignoring status code %u (%s)",
+			 msg->status_code, msg->reason_phrase);
+		return;
+	}
+
+	/* get data */
+	body_length = msg->response_body->length;
+	header_size = soup_message_headers_get_content_length (msg->response_headers);
+	if (header_size < body_length)
+		return;
+
+	/* calculate percentage */
+	percentage = (guint) ((100 * body_length) / header_size);
+	g_debug ("progress: %u%%", percentage);
+	fwupd_client_set_status (client, FWUPD_STATUS_DOWNLOADING);
+	fwupd_client_set_percentage (client, percentage);
+}
+
+/**
+ * fwupd_client_download_bytes:
+ * @client: A #FwupdClient
+ * @url: the remote URL
+ *
+ * Downloads data from a remote server. The fwupd_client_set_user_agent() function
+ * should typically be called before this method is used.
```

Comment:
I think it's worth mandating this. Fail the call if agent not set

Reply:
Right, can do.

Reply:
Done.